File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

int solution(int coin, vector<int> cards) {
    int answer = 1;
    int n = cards.size();
    int target = n + 1;
    
    unordered_set<int> have;
    unordered_set<int> pool;
    
    int idx = 0;
    for (; idx < n/3; idx++) {
        have.insert(cards[idx]);
    }
    
    while(idx < n) {
        
        for (int k = 0; k < 2 && idx < n; k++, idx++) {
            pool.insert(cards[idx]);
        }
        
        bool progressed = false;
        
        int a = -1;
        int b = -1;
        for (int x : have) {
            int y = target - x;
            if (have.count(y)) {
                a = x;
                b = y;
                break;
            }
        }
        
        if (a != -1) {
            have.erase(a);
            have.erase(b);
            progressed = true;
        }
        
        
        if (!progressed && coin >= 1) {
            a = b = -1;
            for (int x : have) {
                int y = target - x;
                if (pool.count(y)) {
                    a = x;
                    b = y;
                    break;
                }
            }
            
            if (a != -1) {
                have.erase(a);
                pool.erase(b);
                coin -= 1;
                progressed = true;
            }
        }
        
        if (!progressed && coin >= 2) {
            a = b = -1;
            for (int x : pool) {
                int y = target - x;
                if (pool.count(y)) {
                    a = x;
                    b = y;
                    break;
                }
            }
            
            if (a != -1) {
                pool.erase(a);
                pool.erase(b);
                coin -= 2;
                progressed = true;
            }
        }
        
        if (!progressed) break;
        answer++;
    }
    return answer;
}

File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.java
import java.util.*;
    
class Solution {
    public int solution(int coin, int[] cards) {
        int n = cards.length;
        int target = n + 1;
        
        Set<Integer> have = new HashSet<>();
        Set<Integer> keep = new HashSet<>();
        
        for (int i = 0; i < n/3; i++) {
            have.add(cards[i]);
        }
        
        int idx = n / 3;
        int round = 1;
        
        while (idx < n) {
            keep.add(cards[idx++]);
            keep.add(cards[idx++]);
            
            // have + have
            if (removePair(have, have, target)) {
                round++;
                continue;
            }
            
            // have + keep
            if (coin >= 1 && removePair(have, keep, target)) {
                coin -= 1;
                round++;
                continue;
            }
            
            // keep + keep
            if (coin >= 2 && removePair(keep, keep, target)) {
                coin -= 2;
                round++;
                continue;
            }
            
            break;
        }
        
        return round;
    }
    
    private boolean removePair(Set<Integer> a, Set<Integer> b, int target) {
        for (int x : a) {
            int y = target - x;
            
            if (!b.contains(y)) continue;
            
            a.remove(x);
            b.remove(y);
            return true;
        }
        return false;
    }
}